### PR TITLE
Targeted salt injection and guard checks

### LINKED ISSUE #608: [BUG] Itch.io deployment fails with "Missing production salt" crash due to GitHub Actions cache
**Title**: [BUG] Itch.io deployment fails with "Missing production salt" crash due to GitHub Actions cache

**Godot Version**: 4.5 stable

**OS**: Web Export (via Ubuntu-latest GitHub Actions runner)

**Steps to Reproduce**:

1. Trigger the `.github/workflows/deploy_to_itch.yml` deployment pipeline.
2. The `inject_salt.sh` script successfully replaces the `"CI_INJECT_SALT_HERE"` placeholder in `scripts/core/globals.gd` with the real production secret.
3. The `firebelley/godot-export` action executes with the `cache: true` parameter.
4. Open the newly deployed build on itch.io and inspect the browser developer console.

**Expected Behavior**:

The game should detect the successfully injected production salt, generate a secure SHA-256 encryption key, and boot normally to the main menu.

**Actual Behavior**:

The game crashes instantly on startup. The production security guard triggers because the compiled game still contains the `"CI_INJECT_SALT_HERE"` placeholder, completely ignoring the bash injection step.

The browser console outputs:

```text
[DEBUG] ⚙️ ENV CHECK (is_automated_test): false
ERROR: CRITICAL SECURITY ERROR: Missing production salt.
  at: push_error (core/variant/variant_utility.cpp:1024)
ERROR: CRITICAL SECURITY ERROR: Missing production salt.
  at: crash (core/core_bind.cpp:336)
Error starting Godot: RuntimeError: unreachable

```

**Screenshots**:

Image

**Additional Context**:

This is a collision between our custom CI pipeline and the Godot export cache.

Because `deploy_to_itch.yml` uses `cache: true` for the `godot-export` action, it pulls down a `.godot/` cache from a previous run. Even though our `inject_salt.sh` script successfully mutates the `globals.gd` source file, the Godot engine relies on the cached compiled bytecode during the headless export phase. As a result, the old placeholder string is packaged into the final `.pck` file.

**Proposed Fix**:

Update `.github/workflows/deploy_to_itch.yml` to explicitly disable the export cache so Godot is forced to recompile the freshly injected GDScript:

```yaml
      - name: "Export Godot to Web"
        id: "export"
        uses: "firebelley/godot-export@930577654862a320eef793f399ee911b4479efb9"
        with:
          # ...
          cache: false  # <-- Change this from true to false
```

### PR DESCRIPTION
**Pull request by ikostan**
Limit CI salt replacement to the exact variable assignment and add verification to avoid corrupting security logic. .github/scripts/inject_salt.sh now uses a specific sed pattern to replace only the var salt line and performs two checks: that the assignment was replaced and that the security guard conditional remains unchanged. scripts/core/globals.gd splits the guard literal (CI_INJECT_ + SALT_HERE) so the injector won't match it. A new test (tests/ci/test_salt_injection.py) asserts the variable is updated while the security guard conditional is preserved. This prevents accidental overwrites of critical runtime checks during CI secret injection.

---
name: Default Pull Request Template
about: Suggesting changes to SkyLockAssault
title: ''
labels: ''
assignees: ''
---

## Description

What does this PR do? (e.g., "Fixes player jump physics in level 2" or "Adds
new enemy AI script")

## Related Issue

Closes #ISSUE_NUMBER (if applicable)

## Changes

- [ ] List key changes here (e.g., "Updated Jump.gd to use Godot 4.4's new Tween
  system")
- [ ] Any breaking changes? (e.g., "Deprecated old signal; migrate to new one")

## Testing

- [ ] Ran the game in Godot v4.5 editor—describe what you tested (e.g., "Jump
  works on Win10 with 60 FPS")
- [ ] Any new unit tests added? (Link to test scene if yes)
- [ ] Screenshots/GIFs if UI-related: (Attach below)

## Checklist

- [ ] Code follows Godot style guide (e.g., snake_case for variables)
- [ ] No console errors in editor/output
- [ ] Ready for review!

## Additional Notes

Anything else? (e.g., "Tested on Win10 64-bit; needs Linux validation")

## Summary by Sourcery

Tighten CI salt injection to only update the intended salt variable while enforcing guards that prevent corruption of security checks.

New Features:
- Add verification in the salt injection script to ensure the salt variable is updated and the security guard conditional remains intact.

Enhancements:
- Restrict the CI salt injection script to target only the specific salt variable assignment rather than all placeholder occurrences.
- Adjust the encryption key guard in globals.gd so that its literal placeholder string is not matched by the CI injector.

Tests:
- Refactor and extend CI salt injection tests to cover targeted replacement, preservation of security guard logic, and various edge cases using shared GDScript content helpers.

## Summary by CodeRabbit

* **Tests**
  * Added regression test for security validation process.

* **Chores**
  * Enhanced internal security validation with improved error handling and verification steps.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ikostan/SkyLockAssault/pull/613?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tests/ci/test_salt_injection.py [hunk 1]
```diff
@@ -308,3 +308,43 @@ def test_injection_unset_secret(repo_tmp):
...
+
+    # 1. Verify the variable was actually updated
+    assert (
+        'var salt: String = "my-production-secret"' in content
+    ), "Variable assignment was not updated!"
+
+    # 2. Verify the security guard conditional remained perfectly intact
+    assert (
+        'if salt == "CI_INJECT_SALT_HERE":' in content
+    ), "FATAL: The security guard logic was overwritten by sed!"
```
[the remark this answers is not in the document]
Comment (ikostan) [reply tests/ci/test_salt_injection.py:337]: Won't fix

**Comment by ikostan** [comment]
@sourcery-ai review

File: tests/ci/test_salt_injection.py [hunk 2]
```diff
@@ -292,19 +287,51 @@ def test_injection_unset_secret(repo_tmp):
...
+def test_injection_does_not_corrupt_security_guard(repo_tmp):
+    """
+    Ensures the injection script uses targeted replacement to update
+    the variable declaration without overwriting the literal string
+    used in the security guard conditional logic.
+    """
+    dummy_rel = f"{repo_tmp}/dummy_guard.gd"
+    dummy_abs = Path(PROJECT_ROOT) / dummy_rel
+
+    original_content = (
```
[the remark this answers is not in the document]
Comment (ikostan) [reply tests/ci/test_salt_injection.py:313]: That is a fantastic code review suggestion. Adding these negative tests ensures your CI pipeline will aggressively catch any future accidental deletions or typos in the security guard.

**Comment by ikostan** [comment]
> * In globals.gd the comment says the guard string is split so sed ignores the line, but the string is still a single literal; either actually split the literal (e.g. via concatenation) or adjust the comment to match the current implementation to avoid confusion.


Fix the Stale Comment in globals.gd (Addressing Issue #2)
Sourcery correctly flagged that the code comment still claims the string is split, even though we merged it back into a single literal to accommodate our sniper-rifle script.

Let's update the comment around line 273 of globals.gd to accurately reflect the new CI/CD validation contract.

**Comment by ikostan** [comment]
> * The inject_salt.sh logic is tightly coupled to the exact text `var salt: String = "CI_INJECT_SALT_HERE"` and `if salt == "CI_INJECT_SALT_HERE":`; consider extracting these patterns into a single shared source of truth (or at least a clearly documented constant) so future refactors of globals.gd don’t silently break the CI injection.

Working on it...

**Comment by ikostan** [comment]
@sourcery-ai review

**Comment by SourceryAI** [comment]
Sorry @ikostan, you have reached your weekly rate limit of 500000 diff characters.

Please try again later or [upgrade](https://app.sourcery.ai/login?connection=github&git_namespace_id=20955183&utm_source=github&utm_medium=bot&utm_campaign=rate_limited) to continue using Sourcery
